src/libsystemd/sd-varlink/sd-varlink.c

```diff
@@ -3262,12 +3272,45 @@ static sd_varlink_server* varlink_server_destroy(sd_varlink_server *s) {
         sd_event_unref(s->event);
 
         free(s->description);
+        free(s->vendor);
+        free(s->product);
+        free(s->version);
+        free(s->url);
 
         return mfree(s);
 }
 
 DEFINE_PUBLIC_TRIVIAL_REF_UNREF_FUNC(sd_varlink_server, sd_varlink_server, varlink_server_destroy);
 
+_public_ int sd_varlink_server_set_info(
+                sd_varlink_server *s,
+                const char *vendor,
+                const char *product,
+                const char *version,
+                const char *url) {
+
+        assert_return(s, -EINVAL);
+        assert_return(vendor, -EINVAL);
+        assert_return(product, -EINVAL);
+        assert_return(version, -EINVAL);
+        assert_return(url, -EINVAL);
+
+        _cleanup_free_ char
+                *a = strdup(vendor),
+                *b = strdup(product),
+                *c = strdup(version),
+                *d = strdup(url);
+        if (!a || !b || !c || !d)
+                return log_oom_debug();
+
+        free_and_replace(s->vendor, a);
+        free_and_replace(s->product, b);
+        free_and_replace(s->version, c);
+        free_and_replace(s->url, d);
```

Comment:
Why not to use `free_and_strdup()` ? Then, this can accept `NULL`.

Reply:
yeah, i think too that we should allow nullifying fields

Reply:
Then it wouldn't be atomic on failure. But you're right, it should accept NULL as arguments.